### PR TITLE
ci: enable CodeBuild PR checks to run on multiple AWS accounts

### PR DESCRIPTION
**Pull request by philasmar**
*Issue #, if available:*
DOTNET-6292

*Description of changes:*
* Created new project `AWS.Deploy.Tools.CI` which will deploy a Lambda function to our main testing account. This Lambda function will ping multiple test runner accounts (which we will define) to check for availability to run the CodeBuild testing project.
* Updated the CloudFormation template `ci.template.yml` to create needed roles without needing anything to be manually defined.
* Added step in GitHub workflow `codebuild-ci.yml` to call the Lambda function to retrieve the appropriate IAM Role to assume before kicking off the CodeBuild project from GitHub.

**Note**: The `AWS CodeBuild CI` PR check is expected to fail for this PR as I am modifying the workflow file to use a different infrastructure which does not yet exist. We need to release this change first for the GitHub workflow to work.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: buildtools/ci.template.yml [hunk 1]
```diff
@@ -116,30 +125,13 @@ Resources:
...
-                Effect: Allow
-                Resource:
-                  - !Ref TestRunnerRoleArn
-              - Action:
-                - codebuild:BatchPutTestCases
-                - codebuild:CreateReport
-                - codebuild:CreateReportGroup
-                - codebuild:UpdateReport
-                - codebuild:UpdateReportGroup
+                - '*'
```
Comment (ashovlin) [inline buildtools/ci.template.yml:128]: Could this be more restrictive than `*`/`*`?

https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#grant-least-privilege

**Review by ashovlin — CHANGES_REQUESTED** [review]

File: .github/workflows/codebuild-ci.yml [hunk 2]
```diff
@@ -16,14 +16,31 @@ jobs:
...
+          aws-region: us-west-2
+      - name: Setup .NET Core 6.0
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: 6.0.x
+      - name: Invoke Load Balancer Lambda
+        id: lambda
+        shell: pwsh
+        run: |
+          aws lambda invoke response.json --function-name "${{ secrets.CI_TESTING_LOAD_BALANCER_LAMBDA_NAME }}" --cli-binary-format raw-in-base64-out --payload '{\"Roles\": \"${{ secrets.CI_TEST_RUNNER_ACCOUNT_ROLES }}\", \"ProjectName\": \"${{ secrets.CI_TESTING_CODE_BUILD_PROJECT_NAME }}\", \"Branch\": \"${{ github.sha }}\"}'
```
[the remark this answers is not in the document]
Comment (philasmar) [reply .github/workflows/codebuild-ci.yml:30]: The AWS CLI is natively available on the runner image https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md.
I am using AWS CLI primarily because I didn't need to install additional tooling (AWS Tools for PowerShell). Also, during testing, I did test out AWS Tools for PowerShell and it piping the output of the command to later steps wasn't straightforward so I stuck with AWS CLI.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            if (!batchGetProjectsResponse.Projects.Any())
+                throw new Exception($"Could not find any project with the name '{input.ProjectName}' in account '{callerIdentity.Account}'.");
+            var project = batchGetProjectsResponse.Projects.First();
+
+            var listBuildsForProjectResponse = await codeBuildClient.ListBuildsForProjectAsync(new ListBuildsForProjectRequest { ProjectName = project.Name });
+
+            var runningBuilds = 0;
+            if (listBuildsForProjectResponse.Ids.Any())
+            {
+                var latestBuilds = listBuildsForProjectResponse.Ids.Take(20).ToList();
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:66]: I wouldn't say guaranteed, but its highly likely. I did it this way because the SDK doesn't have an API for us to retrieve the running builds only. I would needs to retrieve all the jobs that exist in a project and loop through them to get the running builds, which is something I didn't want to do.

File: buildtools/ci.template.yml (hunk 1 above)
Comment (philasmar) [reply buildtools/ci.template.yml:128]: The test runner accounts were already using an IAM role that was highly permissive (`*`/`*`) which was created manually instead of using the template. This is needed because the alternative is to go into every single integration test and quantify which permissions are needed per test. Also, anytime we add new tests that could require additional permissions, we'd have to update our build infra, which is something we don't want to do. I fully acknowledge the best practice, however in our use case, it would become an ops nightmare for us to maintain the list of permissions.

**Review by 96malhar — APPROVED** [review]

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                RoleArn = role,
+                RoleSessionName = "DeployToolTestRunner"
+            });
+
+            var stsClient = new AmazonSecurityTokenServiceClient(assumeRoleResponse.Credentials);
+            var callerIdentity = await stsClient.GetCallerIdentityAsync(new GetCallerIdentityRequest());
+
+            var codeBuildClient = new AmazonCodeBuildClient(assumeRoleResponse.Credentials);
+
+            var batchGetProjectsResponse = await codeBuildClient.BatchGetProjectsAsync(new BatchGetProjectsRequest { Names = new List<string> { input.ProjectName } });
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs]: Done

File: buildtools/README.md [hunk 5]
```diff
@@ -1,11 +1,56 @@
...
-2. Copy output `CodeBuildProjectName` & `OidcRole` output variables.
-3. Create `CI_AWS_ROLE_ARN` repository secret with `OidcRole` value and
-   `CI_AWS_CODE_BUILD_PROJECT_NAME` repository secret with `CodeBuildProjectName`
-   value.
-4. Voila!
+The `AWS.Deploy.Tools.CI` project is used to setup the CI/CD PR check between the [AWS.Deploy.Tools](https://github.com/aws/aws-dotnet-deploy) GitHub Repo and the AWS CodeBuild project that will be executed by the GitHub workflow `AWS CodeBuild CI (codebuild-ci.yml)`.
+
+The workflow will be executed starting from `AWS CodeBuild CI (codebuild-ci.yml)` PR check, which will call a Lambda function in an `AWS Account` (which will act as a load balancer) which will ping 1 or many `Test Runner` accounts to check for availability.
+
+## Main Account (Load Balancer)
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/README.md:7]: As discussed with the team, we will move forward without implementing a more automated deployment approach.

**Comment by philasmar** [comment]
> Approved assuming the CodeBuild CI job passes successfully.

The CodeBuild CI is not expected to pass since the new infrastructure is not yet implemented.

File: .github/workflows/codebuild-ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashovlin) [reply .github/workflows/codebuild-ci.yml:30]: Cool, didn't realize it was provided by GitHub.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashovlin) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:93]: Okay, it does default sort order to descending - https://docs.aws.amazon.com/codebuild/latest/APIReference/API_ListBuildsForProject.html#API_ListBuildsForProject_RequestSyntax

Would feel a little safer if we set that explicitly, but not a blocker.

File: buildtools/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashovlin) [reply buildtools/README.md:7]: * Can you summarize the discussion here for future members of the team, those who weren't at the meeting, etc.
* @normj suggested that we should at least deploy the Lambda via the CLI instead of via Visual Studio which should allow us to write out more of the settings ahead of time.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                await codeBuildClient.BatchGetProjectsAsync(
+                    new BatchGetProjectsRequest
+                    {
+                        Names = new List<string> { input.ProjectName }
+                    }
+                );
+
+            if (!batchGetProjectsResponse.Projects.Any())
+            {
+                throw new Exception($"Could not find any project with the name '{input.ProjectName}' in account '{callerIdentity.Account}'.");
```
Comment (normj) [inline buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:77]: If the project doesn't exist in a particular role should we just skip the role. I'm imagining when this function might be used for a collection of repositories and maybe we haven't created the appropriate codebuild job in all of our test accounts.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    /// The <see cref="GetAvailableTestRunner"/> function is responsible for checking available test runner accounts to run the AWS.Deploy.Tools CodeBuild PR check.
+    /// A list of IAM roles, representing the test runner accounts, is passed to the function. 
+    /// These roles are assumed and used to check if the CodeBuild CI project in the test runner account is currently running any jobs.
+    /// </summary>
+    /// <returns>The function will return the IAM role of the account that is not running any CodeBuild CI jobs.</returns>
+    /// <exception cref="ArgumentNullException">If the input passed to the function is invalid.</exception>
+    /// <exception cref="Exception">If no test runner account is available.</exception>
+    /// <exception cref="Exception">If a CodeBuild CI project is not found in the test runner account.</exception>
+    [LambdaFunction(Name = "GetAvailableTestRunner", Policies = "@AWSDeployToolsCIGetAvailableTestRunnerLambdaAssumeRolePolicy")]
+    public async Task<string> GetAvailableTestRunner([FromBody] GetAvailableTestRunnerInput input)
```
Comment (normj) [inline buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:29]: The `[FromBody]` doesn't do anything here because this is not an API Gateway based function.

File: buildtools/README.md [hunk 8]
```diff
@@ -1,11 +1,56 @@
...
+
+We will start by deploying the `Lambda function` to the main testing account. To do this, load up the `AWS.Deploy.Tools.CI` solution in Visual Studio. Make sure the `AWS Visual Studio Toolkit` is installed. Make sure you have AWS credentials with `Admin` access for the main testing account stored in your `credentials` file locally. The Lambda will be deployed using the `AWS Visual Studio Toolkit`.
+* Right-click on the `AWS.Deploy.Tools.CI` project in Visual Studio
```
Comment (normj) [inline buildtools/README.md:10]: Replace these manual steps with how to deploy with the CLI. Also have a already checked in lambda config file that has all of the requires settings for redeployment. You don't have to manually create the S3 bucket you can use the CLI `--resolve-s3 true` switch to take care of creating the bucket. So there wouldn't have to be anything account specific in the config file.

From this bullet point down to the end could be replaced with something like this.
* dotnet tool update -g Amazon.Lambda.Tools
* dotnet lambda deploy-serverless --region us-west-2 --config-file prod-main-lambda-config.json

**Review by normj — CHANGES_REQUESTED** [review]

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs (hunk 6 above)
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:77]: I threw an exception purposefully to indicate that there is an issue with the infrastructure setup instead of failing silently and us having a hard time tracking why the job was not triggered. This Lambda function is reading IAM roles from the GitHub repo secrets. If the secrets have the IAM role, then I think it's safe to assume that the infrastructure was properly setup, since that IAM role is automatically generated by the CF template that I am using to create the Codebuild job as well.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:100]: image
Setting the sort order will result in an error if the project has more than 100 builds. So I'll leave it as is which default to Descending order.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+        if (!roles.Any())
+        {
+            throw new ArgumentNullException(nameof(input.Roles));
+        }
+
+        foreach (var role in roles)
+        {
+            var assumeRoleResponse =
+                await new AmazonSecurityTokenServiceClient().AssumeRoleAsync(
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs]: Done

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                    new AssumeRoleRequest
+                    {
+                        RoleArn = role,
+                        RoleSessionName = "DeployToolTestRunner"
+                    }
+                );
+
+            var stsClient = new AmazonSecurityTokenServiceClient(assumeRoleResponse.Credentials);
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs]: Done

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            var callerIdentity = await stsClient.GetCallerIdentityAsync(new GetCallerIdentityRequest());
+
+            var codeBuildClient = new AmazonCodeBuildClient(assumeRoleResponse.Credentials);
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs]: Done

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs (hunk 7 above)
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs]: Done

File: buildtools/README.md (hunk 8 above)
Comment (philasmar) [reply buildtools/README.md]: Done

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashovlin) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs:100]: Interesting....okay with it as written then...

File: buildtools/README.md [hunk 12]
```diff
@@ -1,11 +1,52 @@
...
+```
+dotnet tool update -g Amazon.Lambda.Tools
+dotnet lambda deploy-serverless --profile <main-testing-account-admin-profile> --region <main-testing-account-region> --config-file aws-lambda-tools-defaults.json  --resolve-s3 true
+```
+
+The `Lambda Function` should now be deployed in the main testing account. Log into the account and take note of the `Lambda Function` name as it will be needed in a later step. Also, check the Resources tab of the created stack an take note of the IAM role `AWSDeployToolsCIGitHubTrustRole` ARN.
+
+## Test Runner Accounts
+
+You can setup 1 or multiple test runner accounts. These are new AWS Accounts over than the main account that containes the Lambda function. These accounts will run the `AWS CodeBuild Project` defined in [ci.template.yml](./ci.template.yml)
```
Comment (ashovlin) [inline buildtools/README.md:21]: nit: over -> other

File: buildtools/README.md [hunk 13]
```diff
@@ -1,11 +1,52 @@
...
+
+The setup needs an OIDC Identity provider to be defined for the Test Runner account. If one exists, take not of the ARN. If not, go to [Identity providers](https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-west-2#/identity_providers) and create one.
```
Comment (ashovlin) [inline buildtools/README.md:23]: nit: not -> note

**Review by ashovlin — APPROVED** [review]

File: buildtools/README.md (hunk 13 above)
Comment (philasmar) [reply buildtools/README.md]: Fixed

**Review by normj — APPROVED** [review]
One small nit comment but I'm fine approving as is.

File: buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            var assumeRoleResponse =
+                await _stsClient.AssumeRoleAsync(
+                    new AssumeRoleRequest
+                    {
+                        RoleArn = role,
+                        RoleSessionName = "DeployToolTestRunner"
+                    }
+                );
+
+            using var stsClient = new AmazonSecurityTokenServiceClient(assumeRoleResponse.Credentials);
```
[the remark this answers is not in the document]
Comment (philasmar) [reply buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs]: Done
